File: Golang/3. Longest Substring Without Repeating Characters.go
// O(n), O(n)
func lengthOfLongestSubstring(s string) int {
	maxLen, start := 0, 0
	seen := map[rune]int{}

	for idx, ch := range s {
		// duplicate
		if _, ok := seen[ch]; ok && start <= seen[ch] {
			start = seen[ch] + 1 // skip a duplicate
		} else {
			maxLen = max(maxLen, idx-start+1)
		}
		seen[ch] = idx // update last seen index
		fmt.Println(start, seen, maxLen, idx-start+1)
	}
	return maxLen
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Input: s = "abcabcbb"
// Output: 3
// Explanation: The answer is "abc", with the length of 3.

File: Python/3. Longest Substring Without Repeating Characters.py
# Sliding window: O(n), O(n)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        res = 0
        l = 0
        for r in range(len(s)):
            while s[r] in char_set:
                char_set.remove(s[l])
                l += 1
            char_set.add(s[r])
            res = max(res, r-l+1)
        return res
